### ISSUE TITLE
Issues splitting Software 2000 "Copy Blockbuster"-protected titles.

### ISSUE THREAD

**Issue by HeroponRikiBestest** [superg/redumper#199]
See http://forum.redump.org/topic/59306/software-2000-copy-blockbusterprotected-games-need-decisionredumps/ for more info.

This is an overly long issue post just so I can have all the information here if necessary. Skip to the tl;dr reply if you don't need this much info.

MoriGM reached out about wanting to submit a verification for `Swing Plus - Total Mindcontrol (Germany)`, but not being able to because of descramble errors, despite having 0 C2 errors. He gets different hashes between redumper, DiC 20241001, and DiC 20211001. However, each version of each program produces consistent hashes. He dumped it on redumper 4 times and got the same hashes every time. Additionally, the hashes he got from DiC 2021 match the existing dump already in the database. The .scm file also has identical hashes between DiC 2021 and DiC 2024, so this appears to purely be an issue with splitting the data after dumping. Also, Mori has confirmed that his disc has the exact same ringcode as the current entry, so it shouldn't be an issue of different pressings/releases/etc.

After looking into it, what seems to be the issue is that a handful of games from developer Software 2000 seem to use some copy protection called Copy Blockbuster that's responsible for similar errors across discs that use it.

Sarami is apparently aware of this issue, or at least was at some point http://forum.redump.org/post/53958/#p53958

Discs that make use of it:
Pizza Syndicate (specifically just Germany and Italy; although, as mentioned in the cohost writeup linked in my forum post, the French release which isn't in redump atm apparently also uses it too):
http://redump.org/disc/40426/
http://redump.org/disc/76708/
Gute Zeiten, schlechte Zeiten Vol. 2 (Germany)
http://redump.org/disc/108209/
Gute Zeiten, schlechte Zeiten - Fun Pack (Germany)
http://redump.org/disc/108206/
Pizza Syndicate - Mission CD - Mehr Biss (Germany)
http://redump.org/disc/97572/
Swing Plus - Total Mindcontrol (Germany)
http://redump.org/disc/97593/

(I'm aware that base Pizza Syndicate shows 0 errors on both discs, but to avoid making this issue longer than it already is, with all the evidence I have from forum posts and the cohost person's dumping experience, I believe that this is likely just a quirk of how 2017 eccEdc logged some things that weren't C2 errors at the time and how that information was further pulled by users when submitting dumps, and that these discs both have the same issues as the others. Unfortunately, we do not have logs for either to confirm.)

From there, I compared MoriGM's dumps against each other. 
Looking at the logs first; for redumper's logs, redumper fails to descramble the last sector of track 1 and the last sector of track 2 due to ECC/EDC errors. Additionally, every sector [other than the last one?] produces an MSF error, with all but the last 149 also producing ECC/EDC errors.
Both DiC logs show similar results, with a few minor differences. DiC 2021 gives mainerror.txt errors on these 3 sectors:

```
LBA[178994, 0x2bb32]: Track[01]: A part of reversed sector. (Not be scrambled)
LBA[178995, 0x2bb33]: Track[01]: Reversed sector. (Not be scrambled)
LBA[190128, 0x2e6b0]: Track[02]: Reversed sector. (Not be scrambled)
```

Reading the EccEdc.txt logs, DiC 2021 outputs that the last 2 sectors of track 1 produce ECC/EDC errors, and that the last sector of track 2 produces both an MSF error and an ECC/EDC error, rather than just an ECC/EDC error like on redumper.

End of track 1:

```
LBA[178991, 0x2bb2f], MSF[39:48:41], mode 1
LBA[178992, 0x2bb30], MSF[39:48:42], mode 1
LBA[178993, 0x2bb31], MSF[39:48:43], mode 1
LBA[178994, 0x2bb32], MSF[39:48:44], mode 1 User data vs. ecc/edc doesn't match
LBA[178995, 0x2bb33], MSF[38:c8:45], mode 1 User data vs. ecc/edc doesn't match
LBA[178996, 0x2bb34], MSF[ff:aa:55], bad msf
LBA[178997, 0x2bb35], MSF[ff:aa:55], bad msf
LBA[178998, 0x2bb36], MSF[ff:aa:55], bad msf
```

When it hits the last 149 sectors of track 2:

```
LBA[189976, 0x2e618], MSF[ff:aa:55], bad msf
LBA[189977, 0x2e619], MSF[ff:aa:55], bad msf
LBA[189978, 0x2e61a], MSF[ff:aa:55], bad msf
LBA[189979, 0x2e61b], MSF[42:15:04], mode 1
LBA[189980, 0x2e61c], MSF[42:15:05], mode 1
LBA[189981, 0x2e61d], MSF[42:15:06], mode 1
```
And, the output at the end:

```
LBA[190126, 0x2e6ae], MSF[42:17:01], mode 1
LBA[190127, 0x2e6af], MSF[42:17:02], mode 1
LBA[190128, 0x2e6b0], MSF[43:97:03], mode 1 User data vs. ecc/edc doesn't match
[ERROR] Number of sector(s) where bad MSF: 10983
	Sector: 178996, 178997, 178998, 178999, 179000, [...], 189976, 189977, 189978, 
[ERROR] Number of sector(s) where user data doesn't match the expected ECC/EDC: 3
	Sector: 178994, 178995, 190128, 
Total errors: 10986
Total warnings: 0
```

Next, DiC 2024 gives descramble errors on these sectors:

```
LBA[178994, 0x2bb32]: Track[01]: A part of reversed
LBA[178995, 0x2bb33]: Track[01]: Invalid MSF - Reversed. Skip descrambling
LBA[178996, 0x2bb34]: Track[02]: Invalid MSF - Other. Skip descrambling
LBA[178997, 0x2bb35]: Track[02]: Invalid MSF - Other. Skip descrambling
[...]
LBA[189977, 0x2e619]: Track[02]: Invalid MSF - Other. Skip descrambling
LBA[189978, 0x2e61a]: Track[02]: Invalid MSF - Other. Skip descrambling
LBA[190128, 0x2e6b0]: Track[02]: Invalid MSF - Reversed. Skip descrambling
```

Reading the EccEdc.txt logs, DIC 2024 finds that only the second to last sector of track 1 has an ecc/edc error, but not the last sector. It then goes on to say that every sector of track 2 until the last 149 produce ecc/edc errors, despite having identified them as invalid MSF in mainError.txt, and thus getting the LBA wrong as well. Finally, it does not find that the last sector of track 2 produces any sort of errors.

End of track 1:
```
LBA[178991, 0x2bb2f], MSF[39:48:41], mode 1
LBA[178992, 0x2bb30], MSF[39:48:42], mode 1
LBA[178993, 0x2bb31], MSF[39:48:43], mode 1
LBA[178994, 0x2bb32], MSF[39:48:44], mode 1 User data vs. ecc/edc doesn't match
LBA[178995, 0x2bb33], MSF[39:48:45], mode 1
LBA[740155, 0xb4b3b], MSF[fe:2a:55], mode 1 User data vs. ecc/edc doesn't match
LBA[740155, 0xb4b3b], MSF[fe:2a:55], mode 1 User data vs. ecc/edc doesn't match
LBA[740155, 0xb4b3b], MSF[fe:2a:55], mode 1 User data vs. ecc/edc doesn't match
```

When it hits the last 149 sectors of track 2:

```
LBA[740155, 0xb4b3b], MSF[fe:2a:55], mode 1 User data vs. ecc/edc doesn't match
LBA[740155, 0xb4b3b], MSF[fe:2a:55], mode 1 User data vs. ecc/edc doesn't match
LBA[740155, 0xb4b3b], MSF[fe:2a:55], mode 1 User data vs. ecc/edc doesn't match
LBA[189979, 0x2e61b], MSF[42:15:04], mode 1
LBA[189980, 0x2e61c], MSF[42:15:05], mode 1
LBA[189981, 0x2e61d], MSF[42:15:06], mode 1
```

And, the output at the end:
```
LBA[190126, 0x2e6ae], MSF[42:17:01], mode 1
LBA[190127, 0x2e6af], MSF[42:17:02], mode 1
LBA[190128, 0x2e6b0], MSF[42:17:03], mode 1
[ERROR] Number of sector(s) where user data doesn't match the expected ECC/EDC: 10984
	Sector: 178994, 740155, 740155, 740155, 740155, [...], 740155, 740155, 740155, 740155, 740155, 
Total errors: 10984
Total warnings: 0
```

Comparing the dumps themselves directly, the final results on the .bin files can be seen.

Track 1:

DiC 2021 (left) vs DiC 2024/redumper (right; DiC 2024 and redumper produce the same hashes for track 1)

![track1](https://github.com/user-attachments/assets/ea6d7bd5-359d-4389-96c1-ac2e903ce5d7)

The only difference is in the last sector, which appears to not have been descrambled on DiC 2021, and on DiC 2024/redumper it appears to have been descrambled into a normal looking sector with 2048 bytes of 0x00 and 288 bytes of ECC/EDC data at the end.

Track 2: 

DiC 2021 (left) vs redumper (right)

![dic2021vsRedumper_track2](https://github.com/user-attachments/assets/6c616ca4-0684-40cd-9116-8129417ade65)

The only differences between DiC 2021 and redumper (as well as DiC 2024, mentioned below) on track 2 seems to be the same thing as on track 1.

DiC 2024 (left) vs redumper (right)

![dic2024vsRedumper_track2](https://github.com/user-attachments/assets/74982571-8689-486e-9730-c6463adb61ff)

DiC 2024 appears to be, aside from the last 149 sectors, entirely un-descrambled, wheras redumper and DiC 2021 displayed sectors filled with 0xAA and following ECC data up until the last 149 sectors.

DiC 2024 (left) does, though, get the same results as redumper (right) on the last sector of track 2.

![dic2024vsRedumper_matchEnd_track2](https://github.com/user-attachments/assets/f62d70a7-6e95-4c9a-89db-4a76d73aa98b)

**Comment by HeroponRikiBestest** [comment]
Tl;dr:

A specific German developer used copy protection on ~5 of their games, which doesn't impede our drives'/dumping programs' ability to dump them, but results in inconsistencies between how our dumping programs split them after dumping. All of the current dumps in redump likely have issues and need to be redumped, I'm not sure which program handles it correctly, and so I wanted to open an issue to get your input.

DiC 2021 disagrees with DiC 2024 and redumper on the last sector of each track.
DiC 2024 disagrees with DiC 2021 and redumper on what to do about the 10983 sectors of combination ECC/EDC errors and bad MSF.

As a result, all 3 consistently produce different results from the same disc.

**Comment by Deterous** [comment]
I believe all of these differences are just whether each sector is scrambled or not. Redumper is specific in which situations it does not descramble. Likely no changes are needed here.
